XiaoW - optimize the request for user tasks to one post request

wang9hu opened the pull request:
# Description
When loading dashboard, app will send a get request for each user for its task and for admin there will be hundreds of requests. This PR integrates all the request into one post request.

## Related PRS (if any):
This frontend PR is related to the [#402](https://github.com/OneCommunityGlobal/HGNRest/pull/402#issue-1751172814) backend PR. backend PR.
To test this backend PR you need to checkout the [#402](https://github.com/OneCommunityGlobal/HGNRest/pull/402#issue-1751172814) backend PR.frontend PR.
…

## Main changes explained:
- collect required userId into one array
- change the requst to post with a body of userIds
…

## How to test:
1. check into current branch as front end
2. check into [#402](https://github.com/OneCommunityGlobal/HGNRest/pull/402#issue-1751172814) backend PR. branch as back end
3. do `npm install` and `npm run start:local` to run this PR locally
4. log as admin user
5. check the developer tool - network, check the request of `.../TimeEntry/users`, this is the request for all the tasks that used to be hundreds of requests
6. see if there is any freezing as before when loading, or any potential bugs caused by this PR.

bienzguoa reviewed (commented):
Hi XiaoW,
It has two pendings from my testing, the first one "data" finished in 6.8 mins, but I think this is resolved in another PR. The second one is still pending after 15 mins. There is no other error.
<img width="1456" alt="Screenshot 2023-06-10 at 8 10 28 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/b454453f-59c3-40ca-996f-979f2b74220e">
<img width="1448" alt="Screenshot 2023-06-10 at 8 11 01 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/efc1e03a-905c-4620-b08d-0f18ddd1315d">
<img width="736" alt="Screenshot 2023-06-10 at 8 17 08 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/da28f67f-fb05-455a-a574-f46aad1c0f6f">

wang9hu commented:
> Hi XiaoW, It has two pendings from my testing, the first one "data" finished in 6.8 mins, but I think this is resolved in another PR. The second one is still pending after 15 mins. There is no other error. <img alt="Screenshot 2023-06-10 at 8 10 28 PM" width="1456" src="https://user-images.githubusercontent.com/92126559/244906506-b454453f-59c3-40ca-996f-979f2b74220e.png"> <img alt="Screenshot 2023-06-10 at 8 11 01 PM" width="1448" src="https://user-images.githubusercontent.com/92126559/244906507-efc1e03a-905c-4620-b08d-0f18ddd1315d.png"> <img alt="Screenshot 2023-06-10 at 8 17 08 PM" width="736" src="https://user-images.githubusercontent.com/92126559/244906604-da28f67f-fb05-455a-a574-f46aad1c0f6f.png">

Thanks, @bienzguoa! I should be more specific, the request I modified is the `.../TimeEntry/users`, which is a replacement for all the individual get requests. I have changed the description.

bienzguoa reviewed (approved):
Looks good! There is no freezing or errors.
<img width="807" alt="Screenshot 2023-06-10 at 8 33 44 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/c38ccc81-e944-40b0-92ad-038f4ad839e0">

Chuehleo reviewed (approved):
Everything working as intended, no freezing when loading.
![1686589007462](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122568562/540b295c-a0fd-4d21-b41e-fad6138e8b8e)
![1686588970874](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122568562/3ef32008-4c50-423f-847e-952d31207d24)

oreoluwadnd reviewed (dismissed):
![Screenshot from 2023-06-12 19-06-52](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/22573947/3a18cfc0-767a-46a4-98a2-87b8976ddbbf)
![Screenshot from 2023-06-12 19-08-00](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/22573947/f348f5cc-d36a-4eaa-9ee1-911062d4ecc6)
The request was optimized, but the response was Cannot POST /api/TimeEntry/users

KurtisIvey reviewed (approved):
Functions exactly as intended. Great Job!

![pr905](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/19482363/22bc997b-1ccc-4fe5-ae19-95e418540419)

eduhorta reviewed (approved):
Working as intended! Nice job! 👍

![Screenshot 2023-06-13 10-24-25](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/56b2d54f-c383-4f22-9ce1-1870eb438a1e)
![Screenshot 2023-06-13 10-24-27](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/5fb73679-085e-47ea-8b27-55b2d25e23c0)

AuroraLovesWhales reviewed (approved):
Hi @wang9hu. I have tested you PR and the optimization is working as intended. Great work! 
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/68ee365c-c4b6-4aaa-80f2-43358d5eb5d2)
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/48ee70f8-ab6b-49d0-9e7e-62fa3e5474fe)

Lucas-E commented:
@oreoluwadnd Please, re-review it so I can move it to final review!

cvtqx reviewed (approved):
Seems to be working as described, no freezing noticed while loading. 

![Screen Shot 2023-06-14 at 3 41 25 PM](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/107726913/19326589-c986-4f46-aa62-f06492036d06)

yufeiStella commented:
Hi @wang9hu. It seems that there are two occurrences of the http://localhost:4500/api/TimeEntry/users. One is a successful POST method, while the other is an OPTIONS request method, and it returns a Status Code of 204 No Content.

POST: 
<img width="766" alt="Screen Shot 2023-06-15 at 15 32 41" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/1361ef76-4e7c-4a48-8008-819da959e942">
<img width="972" alt="Screen Shot 2023-06-15 at 15 36 07" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/daecc4d6-e7ad-4068-8f65-e56e3622f328">

OPTIONS:
<img width="745" alt="204" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/383e993f-4fb2-4bbd-9306-5972faa57184">

<img width="877" alt="Screen Shot 2023-06-15 at 15 39 23" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111897624/a6c93e63-a88e-4669-9565-c919d1867d21">

wang9hu commented:
> Hi @wang9hu. It seems that there are two occurrences of the http://localhost:4500/api/TimeEntry/users. One is a successful POST method, while the other is an OPTIONS request method, and it returns a Status Code of 204 No Content.

Hi @yufeiStella , thanks for reviewing my PR. That OPTION request is a preflight request sent by CORS, I am not quite familiar with that, please see details in [Preflight request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request#:%7E:text=It%20is%20an%20OPTIONS%20request,to%20craft%20such%20requests%20themselves.)

EdwinLauM reviewed (approved):
Works as intended! Good job! 

<img width="1676" alt="Screen Shot 2023-06-16 at 6 27 02 AM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/8802ddc1-4b0d-4c09-b49b-7a0df1dab844">

one-community commented:
Thank you all, moving this to final review!

OAkinbode reviewed (approved):
Hello XiaoW, The PR works as intended. Much faster loading time also. Well done. 
<img width="1440" alt="PR 905 working" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/aa376fbd-acdf-4721-939f-da85d88884e8">

AuroraLovesWhales reviewed (approved):
Reviewed this PR again and everything is working as intended. Great job!
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/8947c62f-ecc2-4289-ae83-d0a439d46d66)

KaixiangGU commented:
Reviewed, great job on optimizing the api call!
![Capture](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/98206107/01855f81-4f65-4ef3-969a-bd062b718f8a)

mahamakhan reviewed (approved):
Checked the backend too works fine for me

<img width="1044" alt="image" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/106139234/36a3bcd8-c0b8-4ed0-84d4-973628c536f4">
